### ISSUE TITLE
Type of 'onClick' regression

### ISSUE THREAD

**Issue by ShimiSun** [grommet/grommet#7016]
_Originally posted by @ShimiSun in https://github.com/grommet/grommet/issues/6983#issuecomment-1810580064_
            
Our team is experiencing a backward compatibility issue due to this fix. This PR is limiting the range of mouse event type and we are seeing the following type error:
```
 Type '((event: MouseEvent<HTMLDivElement, MouseEvent>) => void) | undefined' is not assignable to type 
 '((event: MouseEvent<Element, MouseEvent>) => void) | undefined'.
 ```
 Hence, until now, grommet accepted `HTMLDivElement`, but now it only allows `Element`. I suspect this may not be backward compatible. https://developer.mozilla.org/en-US/docs/Web/API/HTMLDivElement

Here are a couple (out of 8)  of workarounds we had to apply in order to get the new typing to work
![image](https://user-images.githubusercontent.com/6320236/282874809-acad1111-11f9-48a0-adc5-1eb57c7c050d.png)

Unfortunately, this is reducing our type-specific and making our code less safe.

**Comment by akellbl4** [comment]
A proper solution for this problem would be next taking in type of a tag name that box renders and building type of `onClick` by taking it from corresponding html element to that tag.

The problem that this types have now. When we define `React.MouseEvent` as event that `Box` would produce `onClick` it's to common. On the opposite side when we say that it produces event with type `React.MouseEvent<HTMLDivElement>` breaking types for cases when `as` or `tag` is used, because if you have `<Box as="button" onClick={(evt) => {...}}/>` you should receive event that is `React.MouseEvent<HTMLButtonElement>`. So you can try to access some specific thing of HTMLButtonElement in event handler and typescript will scream at you that HTMLDivElement doesn't have that property.

I've created an example of how it can be [done here](https://www.typescriptlang.org/play?pretty=false#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wG4AocmATzCTgEE00kwYBnAFRQHN24AvHADWSahExwAUgGUAGgDoAkgDsYUYCvbA0AUQA2SEEjXsKVWvQAKOMOwA8nOEgAeMEwBN+TFmy692AD5Bcjg4AHpwuG0VFjg0XEgVE3gACxR+AAMMzLgUFQ84TJheXIB3emSkQpgIOCJ2dR14FAAjYH1gGjhauDAM-laIGFS4Mq7Rj2B2NA0QTXz4GjpQiKiy1J1RkZwy-iQoHCgx1JM4cQBXPoG4IZHVgAoAbzz2AH4ALjhOUh7eT7gyQAbgc4ABfOAAHzgLxKPABP1eAOBoLBAEo4AAyGGrMIgag2CBgAGNDQqHgUMIQ7FWHTCezIdAwBQAYUSEGSakJdkcgQANAQOSzOmhhPgoQQ0Jt9B4iCp8IFzJgLrEYMAOXAAEIQFyOZxuTzeZisDjcPiCAhTIEKh5gWzsL7chycQIYp6rIgwC5QFRweyBfLUEaaHj2cKK8hgyiuSCwODK1Xq31stQoTQHB5u1YJLRpfIeQzCukAeRUABFgECvoyMAoALIQC7sJC6EFqAAS+cMUHs7c4dYAMhWgQYjClgkIHkggTAMQJgu6wmEc+wIIYFPoIDwHvg0CLhHArfgBdPZ5Sl5El1frzebwA9N6rMHmZccxpwdIFQv70syfoqatiGZBsmxbNsYE7L8Dl7fsBz-fJR2MNQJzgKcZznBdcXiN81yQDctx3Pc6UPStj2cdDzzCS9bxomiHyfF96iQL0fVQrD-SwsJ7G1Fw4CFfcBCeT8CyQItRVLYcwUCK0wx4wJOL9HjXgEfB2H-cV+LpQThO-EsVHglQpIUsI1PyBTZJ1eSbyecIACpnEOaBBiQNAUFAviVH0agPPoCQelSCBmzgaY8n0TcKhqOpWnofp2GbQpbPCKMb24nVlNU9S-h4FTWguGBanlQILJcKzr1S3i4RyvKCvFDIVNMwritKrjw1WNEKCjIA)

In few words it works like that:
- turn type of tag into generic and only `tag` or `as` prop to be passed so you couldn't pass 2 different tags in props (line 8)
- by default `Box` rendered as `div` so we take `onClick` type from it (line 12)
- if `tag` or `as` passed `onClick` will be taken from props of those "components" (fyi pure tags are primitive components of React)

With these changes types will work correctly with all of the variations of props that can be introduced.

**Comment by akellbl4** [comment]
cc @jcfilben

**Comment by jcfilben** [comment]
Resolved by #7018
